File: sdks/sandbox/go/manager.go
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package opensandbox

import (
	"context"
	"time"
)

// SandboxManager provides administrative operations on sandboxes
// without connecting to a specific sandbox.
type SandboxManager struct {
	lifecycle *LifecycleClient
}

// NewSandboxManager creates a SandboxManager from the given connection config.
func NewSandboxManager(config ConnectionConfig) *SandboxManager {
	return &SandboxManager{
		lifecycle: config.lifecycleClient(),
	}
}

// ListSandboxInfos returns a paginated list of sandboxes with optional filtering.
func (m *SandboxManager) ListSandboxInfos(ctx context.Context, filter ListOptions) (*ListSandboxesResponse, error) {
	return m.lifecycle.ListSandboxes(ctx, filter)
}

// GetSandboxInfo retrieves info for a single sandbox by ID.
func (m *SandboxManager) GetSandboxInfo(ctx context.Context, sandboxID string) (*SandboxInfo, error) {
	return m.lifecycle.GetSandbox(ctx, sandboxID)
}

// PatchSandboxMetadata patches metadata for a sandbox by ID.
func (m *SandboxManager) PatchSandboxMetadata(ctx context.Context, sandboxID string, patch MetadataPatch) (*SandboxInfo, error) {
	return m.lifecycle.PatchSandboxMetadata(ctx, sandboxID, patch)
}

// KillSandbox terminates a sandbox by ID.
func (m *SandboxManager) KillSandbox(ctx context.Context, sandboxID string) error {
	return m.lifecycle.DeleteSandbox(ctx, sandboxID)
}

// PauseSandbox pauses a running sandbox by ID.
func (m *SandboxManager) PauseSandbox(ctx context.Context, sandboxID string) error {
	return m.lifecycle.PauseSandbox(ctx, sandboxID)
}

// ResumeSandbox resumes a paused sandbox by ID.
func (m *SandboxManager) ResumeSandbox(ctx context.Context, sandboxID string) error {
	return m.lifecycle.ResumeSandbox(ctx, sandboxID)
}

// RenewSandbox extends a sandbox's expiration by the given duration from now.
func (m *SandboxManager) RenewSandbox(ctx context.Context, sandboxID string, duration time.Duration) (*RenewExpirationResponse, error) {
	return m.lifecycle.RenewExpiration(ctx, sandboxID, time.Now().Add(duration))
}

// CreateSnapshot creates a snapshot for the given sandbox.
func (m *SandboxManager) CreateSnapshot(ctx context.Context, sandboxID string, req CreateSnapshotRequest) (*SnapshotInfo, error) {
	return m.lifecycle.CreateSnapshot(ctx, sandboxID, req)
}

// GetSnapshot retrieves snapshot info by ID.
func (m *SandboxManager) GetSnapshot(ctx context.Context, snapshotID string) (*SnapshotInfo, error) {
	return m.lifecycle.GetSnapshot(ctx, snapshotID)
}

// ListSnapshots returns a paginated list of snapshots with optional filtering.
func (m *SandboxManager) ListSnapshots(ctx context.Context, filter ListSnapshotsOptions) (*ListSnapshotsResponse, error) {
	return m.lifecycle.ListSnapshots(ctx, filter)
}

// DeleteSnapshot deletes a snapshot by ID.
func (m *SandboxManager) DeleteSnapshot(ctx context.Context, snapshotID string) error {
	return m.lifecycle.DeleteSnapshot(ctx, snapshotID)
}

// CreateTemplate declares a fsb template (golden-image build).
// The build is asynchronous: the response starts at TemplatePhasePending;
// poll GetTemplate until the phase reaches TemplatePhaseSucceeded.
func (m *SandboxManager) CreateTemplate(ctx context.Context, req CreateTemplateRequest) (*TemplateInfo, error) {
	return m.lifecycle.CreateTemplate(ctx, req)
}

// GetTemplate retrieves a template with its latest build status by ID.
func (m *SandboxManager) GetTemplate(ctx context.Context, templateID string) (*TemplateInfo, error) {
	return m.lifecycle.GetTemplate(ctx, templateID)
}

// ListTemplates returns a paginated list of templates with optional filtering.
func (m *SandboxManager) ListTemplates(ctx context.Context, opts ListTemplatesOptions) (*ListTemplatesResponse, error) {
	return m.lifecycle.ListTemplates(ctx, opts)
}

// DeleteTemplate deletes a template by ID. Sandboxes already created from the
// template are unaffected.
func (m *SandboxManager) DeleteTemplate(ctx context.Context, templateID string) error {
	return m.lifecycle.DeleteTemplate(ctx, templateID)
}

// Close releases local resources. Currently a no-op placeholder.
func (m *SandboxManager) Close() error { return nil }

File: sdks/sandbox/javascript/src/manager.ts
// Copyright 2026 Alibaba Group Holding Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { ConnectionConfig, type ConnectionConfigOptions } from "./config/connection.js";
import { createDefaultAdapterFactory } from "./factory/defaultAdapterFactory.js";
import type { AdapterFactory } from "./factory/adapterFactory.js";

import type {
  CreateSnapshotRequest,
  ListSandboxesResponse,
  ListSnapshotsParams,
  ListSnapshotsResponse,
  SandboxId,
  SandboxInfo,
  SandboxMetadataPatch,
  SnapshotInfo,
} from "./models/sandboxes.js";
import type {
  CreateTemplateRequest,
  ListTemplatesParams,
  ListTemplatesResponse,
  TemplateInfo,
} from "./models/templates.js";
import type { Sandboxes } from "./services/sandboxes.js";

export interface SandboxManagerOptions {
  /**
   * Connection configuration for calling the OpenSandbox Lifecycle API.
   */
  connectionConfig?: ConnectionConfig | ConnectionConfigOptions;
  /**
   * Advanced override: inject a custom adapter factory (custom transports, dependency injection).
   */
  adapterFactory?: AdapterFactory;
}

export interface SandboxFilter {
  /**
   * Filter by sandbox lifecycle states.
   */
  states?: string[];
  /**
   * Filter by metadata key-value pairs.
   */
  metadata?: Record<string, string>;
  /**
   * Pagination page number (1-indexed).
   */
  page?: number;
  /**
   * Number of items per page.
   */
  pageSize?: number;
}

/**
 * Administrative interface for managing sandboxes (list/get/pause/resume/kill/renew).
 *
 * For interacting *inside* a sandbox, use {@link Sandbox}.
 */
export class SandboxManager {
  private readonly sandboxes: Sandboxes;
  private readonly connectionConfig: ConnectionConfig;

  private constructor(opts: { sandboxes: Sandboxes; connectionConfig: ConnectionConfig }) {
    this.sandboxes = opts.sandboxes;
    this.connectionConfig = opts.connectionConfig;
  }

  static create(opts: SandboxManagerOptions = {}): SandboxManager {
    const baseConnectionConfig = opts.connectionConfig instanceof ConnectionConfig
      ? opts.connectionConfig
      : new ConnectionConfig(opts.connectionConfig);
    const connectionConfig = baseConnectionConfig.withTransportIfMissing();
    const lifecycleBaseUrl = connectionConfig.getBaseUrl();
    const adapterFactory = opts.adapterFactory ?? createDefaultAdapterFactory();
    let sandboxes: Sandboxes;
    try {
      sandboxes = adapterFactory.createLifecycleStack({
        connectionConfig,
        lifecycleBaseUrl,
      }).sandboxes;
    } catch (err) {
      void connectionConfig.closeTransport().catch(() => undefined);
      throw err;
    }
    return new SandboxManager({ sandboxes, connectionConfig });
  }

  listSandboxInfos(filter: SandboxFilter = {}): Promise<ListSandboxesResponse> {
    return this.sandboxes.listSandboxes({
      states: filter.states,
      metadata: filter.metadata,
      page: filter.page,
      pageSize: filter.pageSize,
    });
  }

  getSandboxInfo(sandboxId: SandboxId): Promise<SandboxInfo> {
    return this.sandboxes.getSandbox(sandboxId);
  }

  patchSandboxMetadata(
    sandboxId: SandboxId,
    patch: SandboxMetadataPatch,
  ): Promise<SandboxInfo> {
    return this.sandboxes.patchSandboxMetadata(sandboxId, patch);
  }

  killSandbox(sandboxId: SandboxId, signal?: AbortSignal): Promise<void> {
    return this.sandboxes.deleteSandbox(sandboxId, signal);
  }

  pauseSandbox(sandboxId: SandboxId): Promise<void> {
    return this.sandboxes.pauseSandbox(sandboxId);
  }

  resumeSandbox(sandboxId: SandboxId): Promise<void> {
    return this.sandboxes.resumeSandbox(sandboxId);
  }

  /**
   * Renew expiration by setting expiresAt to now + timeoutSeconds.
   */
  async renewSandbox(sandboxId: SandboxId, timeoutSeconds: number): Promise<void> {
    const expiresAt = new Date(Date.now() + timeoutSeconds * 1000).toISOString();
    await this.sandboxes.renewSandboxExpiration(sandboxId, { expiresAt });
  }

  createSnapshot(sandboxId: SandboxId, req?: CreateSnapshotRequest): Promise<SnapshotInfo> {
    return this.sandboxes.createSnapshot(sandboxId, req);
  }

  getSnapshot(snapshotId: string): Promise<SnapshotInfo> {
    return this.sandboxes.getSnapshot(snapshotId);
  }

  listSnapshots(filter: ListSnapshotsParams = {}): Promise<ListSnapshotsResponse> {
    return this.sandboxes.listSnapshots(filter);
  }

  deleteSnapshot(snapshotId: string): Promise<void> {
    return this.sandboxes.deleteSnapshot(snapshotId);
  }

  /**
   * Create a fsb template (golden-image build).
   *
   * The build is asynchronous: the response starts at `status.phase: Pending`;
   * poll `getTemplate` until `Succeeded` (or `Failed`).
   */
  createTemplate(req: CreateTemplateRequest): Promise<TemplateInfo> {
    return this.sandboxes.createTemplate(req);
  }

  /**
   * Get a template with its latest build status by id.
   */
  getTemplate(templateId: string): Promise<TemplateInfo> {
    return this.sandboxes.getTemplate(templateId);
  }

  /**
   * List templates with metadata filtering and pagination options.
   */
  listTemplates(filter: ListTemplatesParams = {}): Promise<ListTemplatesResponse> {
    return this.sandboxes.listTemplates(filter);
  }

  /**
   * Delete a template by id. Running sandboxes are unaffected.
   */
  deleteTemplate(templateId: string): Promise<void> {
    return this.sandboxes.deleteTemplate(templateId);
  }

  /**
   * Release the HTTP agent resources allocated for this manager instance.
   *
   * Each manager clone owns a scoped `ConnectionConfig` clone.
   *
   * This mirrors the Python SDK's default transport lifecycle.
   */
  async close(): Promise<void> {
    await this.connectionConfig.closeTransport();
  }
}
